### DISCUSSION TITLE
Opus codec?

### DISCUSSION THREAD

**Discussion by dholth** [micropython/micropython#10756] [Libraries & Drivers]
I thought it would be interesting to get the Opus codec into micropython. Some people run it on the ESP32 for example. [This handy distribution](https://github.com/pschatzmann/arduino-libopus) is [easy to compile with cmake,](https://github.com/dholth/micropython/tree/opus/examples/opusmodule/opus) but maybe not for the Unix port. Branch at https://github.com/dholth/micropython/tree/opus/examples/usercmodule/opus. It looks like I need to edit the `micropython.mk` to match the `micropython.cmake` if I want the other build to work; read through all of the included C modules to figure out what the API is; ...

**Comment by jimmo** [comment]
See also parallel discussion at https://discordapp.com/channels/574275045187125269/1012726713291128862/1075251278009073724

**Comment by dholth** [comment]
I was able to get this to compile and have called my first opus function, on the Unix port. To avoid undefined-QSTR errors, I had to avoid compiling the existing cexample and cppexample, and change my sources to `SRC_USERMOD_C` instead of `SRC_USERMOD`. On Ubuntu it requires `apt install libopusfile-dev`, and [a script](https://github.com/dholth/micropython/blob/opus/ports/unix/mopus.sh) passes the necessary `make` arguments.

**Comment by dholth** [comment]
I was able to decode opus in micropython running on Unix.

```
import opus
music = open("some-opus-file.ogg", "rb")
op = opus.opus(music)
buffer = memoryview(bytearray(23040))
output = open("decoded.raw", "w+b")
min_samples = 30000
max_samples = 0
while samples := op.read_stereo(buffer):
    min_samples = min(samples, min_samples)
    max_samples = max(samples, max_samples)
    output.write(buffer[:samples*4])
```

(Even though they recommend a big buffer, the file I tried pulls 960-samples or 3840-bytes per call maximum)

**Comment by dholth** [comment]
Unfortunately, libopusfile is a little less friendly to cross-compiled cmake compared to ogg/opus, or arduino-libopus...

**Comment by dholth** [comment]
I was able to get this to work under [CircuitPython](https://github.com/dholth/micropython/tree/opus-cp/examples/oggzmodule/oggz) after addressing some minor extension module differences, tweaking [arduino-libopus](https://github.com/dholth/arduino-libopus) and writing a makefile. Still pretty rough, and doesn't use the C-only smooth audio playback API for circuitpython. Need to keep at it until we can feed it data from a network; micropython/circuitpython doesn't think "a Python class with a read() method" is a file.

**Comment by dholth** [comment]
I was able to get Opus decoding to work under MicroPython as an external module, in https://github.com/dholth/arduino-libopus/blob/main/micropython/opusdec_i2s.py.

I can easily fit 2 1/2 minutes of 48kbps stereo audio as a 1MB file on the RP2040's built-in flash.

It compiles something like this `make -j BOARD=PICO USER_C_MODULES=$HOME/prog/arduino-libopus/micropython/micropython.cmake CFLAGS_EXTRA=-DMODULE_OGGZ_ENABLED=1 V=1`

Unlike CircuitPython, when compiled under micropython we don't have to hack in the following symbols:

```
_fstat;
__gnu_thumb1_case_shi;
_close;
_isatty;
_lseek;
```

I'm playing the audio through a MAX98357A DAC which should expect right-justified samples, which I understand is what machine.I2S spits out.

`audio_out = I2S(0, sck=Pin(SCK_PIN), ws=Pin(WS_PIN), sd=Pin(SD_PIN), mode=I2S.TX, bits=16, format=(stereo), rate=48000, ibuf=BUFFER_LENGTH_IN_BYTES)`

Playback quality is pretty good with minor gitches, as long as I use the blocking API and set ibuf=BUFFER_LENGTH_IN_BYTES to the same 20ms chunks that the Opus codec produces. Although it sounds like it is clipping, and I haven't gotten the I2S.shift() function working yet to reduce the volume.

The IRQ API is extremely glitchy, at least how I'm using it. I tried single and double buffering the decoded samples, either decoding it right away or decoding the next callback's samples after writing the current ones. The IRQ callback is passed an undocumented parameter which from the C code appears to be the I2S object.

I have not tried the async API.

**Reply by lexiforest** [reply]
Cool!, are there any updates on your work?

**Reply by dholth** [reply]
I didn't quite figure out the API that micropython or circuitpython (different) expects for smooth audio playback, but it would be awesome to continue.
